doc/connectivity/usb/device/usb_device.rst

```diff
@@ -447,6 +447,70 @@ before transmitting any data.
 
 .. _testing_USB_native_sim:
 
+Interface number and endpoint address assignment
+************************************************
+
+In USB terminology, a ``function`` is a device that provides a capability to the
+host, such as a HID class device that implements a keyboard. A function
+constains a collection of ``interfaces``; at least one interface is required. An
+interface may contain device ``endpoints``; for example, at least one input
+endpoint is required to implement a HID class device, and no endpoints are
+required to implement a USB DFU class. A USB device that combines functions is
+a multifunction USB device, for example, a combination of a HID class device
+and a CDC ACM device.
+
+With Zephyr RTOS USB support, various combinations are possible with built-in USB
+classes/functions or custom user implementations. The limitation is the number
+of available device endpoints. Each device endpoint is uniquely addressable.
+The endpoint address is a combination of endpoint direction and endpoint
+number, a four-bit value. Endpoint number zero is used for the default control
+method to initialize and configure a USB device. By specification, a maximum of
+``15 IN`` and ``15 OUT`` device endpoints are also available for use in functions.
+The actual number depends on the device controller used. Not all controllers
+support the maximum number of endpoints and all endpoint types. For example, a
+device controller might support one IN and one OUT isochronous endpoint, but
+only for endpoint number 8, resulting in endpoint addresses 0x88 and 0x08.
+Also, one controller may be able to have IN/OUT endpoints on the same endpoint
+number, interrupt IN endpoint 0x81 and bulk OUT endpoint 0x01, while the other
+may only be able to handle one endpoint per endpoint number. Information about
+the number of interfaces, interface associations, endpoint types, and addresses
+is provided to the host by the interface, interface specifiec, and endpoint
+descriptors.
```

Comment:
A common misinterpretation I have observed is that people tend to consider 0x81 and 0x01 as a single endpoint working in both directions. However, in reality, any non-control endpoint is always unidirectional, meaning that 0x81 and 0x01 are completely independent endpoints just like 0x82 from 0x01 is. Not sure how to concisely write it.

Also while specification allows non-zero control endpoints, I don't think Zephyr supports it.

Reply:
Maybe some controller documentation is to blame, and our binding description does not make it better. e.g. from STM32 RM0431 "Device bidirectional endpoints (including EP0)" ended up in dts/bindings/usb/usb-ep.yaml


```
      Number of bi-directional endpoints supported by hardware
      (including EP0)
```